jmh-benchmarks/src/main/java/org/apache/kafka/jmh/streams/MockMemoryLRUCache.java

```diff
@@ -0,0 +1,275 @@
+ */
+package org.apache.kafka.jmh.streams;
+
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateRestoreCallback;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StateSerdes;
+import org.apache.kafka.streams.state.internals.InMemoryKeyValueLoggedStore;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An in-memory LRU cache store based on HashSet and HashMap.
+ *
+ * This class has been modified from the original
+ * class to enable testing of caching objects by memory.
+ *
+ *  * Note that the use of array-typed keys is discouraged because they result in incorrect ordering behavior.
+ * If you intend to work on byte arrays as key, for example, you may want to wrap them with the {@code Bytes} class,
+ * i.e. use {@code RocksDBStore<Bytes, ...>} rather than {@code RocksDBStore<byte[], ...>}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ *
+ * @see org.apache.kafka.streams.state.Stores#create(String)
+ */
+public class MockMemoryLRUCache<K, V> implements KeyValueStore<K, V> {
+
+    public interface EldestEntryRemovalListener<K, V> {
+
+        void apply(K key, V value);
+    }
+    private final Serde<K> keySerde;
+
+    private final Serde<V> valueSerde;
+    private String name;
+    protected Map<K, V> map;
+    protected Map<Bytes, byte[]> bytesMap;
+    private StateSerdes<K, V> serdes;
+    private volatile boolean open = true;
+    private boolean isMaxCacheByMemory = false;
+    private boolean isMeasureDeep = false;
+    private boolean isCachingBytes = false;
+    private int totalMemoryUsed = 0;
+    private org.github.jamm.MemoryMeter memoryMeter = new org.github.jamm.MemoryMeter().withGuessing(org.github.jamm.MemoryMeter.Guess.FALLBACK_UNSAFE);
+
+    protected EldestEntryRemovalListener<K, V> listener;
+    protected EldestEntryRemovalListener<Bytes, byte[]> bytesCacheListener;
+
+    // this is used for extended MemoryNavigableLRUCache only
+    public MockMemoryLRUCache(Serde<K> keySerde, Serde<V> valueSerde) {
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+    }
+
+    public MockMemoryLRUCache(String name, final int maxCacheSize, Serde<K> keySerde, Serde<V> valueSerde) {
+        this(keySerde, valueSerde);
+        this.name = name;
+
+        // leave room for one extra entry to handle adding an entry before the oldest can be removed
+        this.map = new LinkedHashMap<K, V>(maxCacheSize + 1, 1.01f, true) {
+            private static final long serialVersionUID = 1L;
+
+            //just for testing would limit cache size to ~ 2GB
+            @Override
+            public int size() {
```

Comment:
Curious as to why you override `size`? Can't you just use `totalMemoryUsed` in  `removeEldestEntry`? Just seems a bit strange to morph `size` into something other than its intended meaning.